File: python/tdigest.py
"""Dunning t-digest for online quantile estimation.

Merging digest variant with K_1 (arcsine) scale function.
Pure Python -- only uses the math standard library.

Uses an array-backed 2-3-4 tree with four-component monoidal measures
for O(log n) quantile lookups, CDF queries, and neighbor finding.
"""

import math
from dataclasses import dataclass

from tree234 import Tree234


@dataclass
class Centroid:
    mean: float
    weight: float


class TdMeasure:
    """Four-component monoidal measure for t-digest centroids."""
    __slots__ = ['weight', 'count', 'max_mean', 'mean_weight_sum']

    def __init__(self, w=0.0, c=0, mm=float('-inf'), mws=0.0):
        self.weight = w
        self.count = c
        self.max_mean = mm
        self.mean_weight_sum = mws


def _centroid_measure(c):
    """Measure for a single centroid key."""
    return TdMeasure(c.weight, 1, c.mean, c.mean * c.weight)


def _combine(a, b):
    """Monoidal combine of two measures."""
    return TdMeasure(
        a.weight + b.weight,
        a.count + b.count,
        max(a.max_mean, b.max_mean),
        a.mean_weight_sum + b.mean_weight_sum,
    )


def _identity():
    """Monoidal identity."""
    return TdMeasure()


def _compare_centroids(a, b):
    """Compare centroids by mean, breaking ties by weight then id."""
    if a.mean < b.mean:
        return -1
    if a.mean > b.mean:
        return 1
    # Tie-break by weight for stability
    if a.weight < b.weight:
        return -1
    if a.weight > b.weight:
        return 1
    return 0


def _weight_of(measure):
    """Extract weight from a TdMeasure."""
    return measure.weight


class TDigest:
    DEFAULT_DELTA = 100
    BUFFER_FACTOR = 5

    def __init__(self, delta: float = DEFAULT_DELTA) -> None:
        self.delta = float(delta)
        self._tree = Tree234(_centroid_measure, _combine, _identity, _compare_centroids)
        self._buffer: list[Centroid] = []
        self.total_weight = 0.0
        self.min = math.inf
        self.max = -math.inf
        self._buffer_cap = math.ceil(self.delta * self.BUFFER_FACTOR)

    # -- mutation ---------------------------------------------------------

    def add(self, value: float, weight: float = 1.0) -> "TDigest":
        value = float(value)
        weight = float(weight)
        self._buffer.append(Centroid(value, weight))
        self.total_weight += weight
        if value < self.min:
            self.min = value
        if value > self.max:
            self.max = value
        if len(self._buffer) >= self._buffer_cap:
            self.compress()
        return self

    def compress(self) -> "TDigest":
        if not self._buffer and self._tree.size() <= 1:
            return self

        all_c = self._tree.to_list() + self._buffer
        self._buffer = []
        all_c.sort(key=lambda c: c.mean)

        new = [Centroid(all_c[0].mean, all_c[0].weight)]
        weight_so_far = 0.0
        n = self.total_weight

        for i in range(1, len(all_c)):
            proposed = new[-1].weight + all_c[i].weight
            q0 = weight_so_far / n
            q1 = (weight_so_far + proposed) / n

            if proposed <= 1 and len(all_c) > 1:
                self._merge_into_last(new, all_c[i])
            elif self._k(q1) - self._k(q0) <= 1.0:
                self._merge_into_last(new, all_c[i])
            else:
                weight_so_far += new[-1].weight
                new.append(Centroid(all_c[i].mean, all_c[i].weight))

        self._tree.clear()
        for c in new:
            self._tree.insert(c)
        return self

    def merge(self, other: "TDigest") -> "TDigest":
        for c in other._flush_for_merge():
            self.add(c.mean, c.weight)
        return self

    # -- queries ----------------------------------------------------------

    def quantile(self, q: float) -> float | None:
        if self._buffer:
            self.compress()
        count = self._tree.size()
        if count == 0:
            return None

        centroids = self._tree.to_list()
        if count == 1:
            return centroids[0].mean

        q = max(0.0, min(1.0, q))
        n = self.total_weight
        target = q * n

        # Use find_by_weight for O(log n) lookup
        result = self._tree.find_by_weight(target, _weight_of)
        if result is None:
            return None

        c, cumulative, i = result

        if i == 0:
            if target < c.weight / 2.0:
                if c.weight == 1:
                    return self.min
                return self.min + (c.mean - self.min) * (target / (c.weight / 2.0))

        if i == count - 1:
            if target > n - c.weight / 2.0:
                if c.weight == 1:
                    return self.max
                remaining = n - c.weight / 2.0
                return c.mean + (self.max - c.mean) * ((target - remaining) / (c.weight / 2.0))
            return c.mean

        mid = cumulative + c.weight / 2.0
        # Need next centroid - get it from centroids list
        next_c = centroids[i + 1]
        next_mid = cumulative + c.weight + next_c.weight / 2.0

        if target <= next_mid:
            frac = 0.5 if next_mid == mid else (target - mid) / (next_mid - mid)
            return c.mean + frac * (next_c.mean - c.mean)

        return self.max

    def cdf(self, x: float) -> float | None:
        if self._buffer:
            self.compress()
        count = self._tree.size()
        if count == 0:
            return None
        if x <= self.min:
            return 0.0
        if x >= self.max:
            return 1.0

        n = self.total_weight
        centroids = self._tree.to_list()

        # Binary search for position among centroid means
        # Find leftmost centroid whose mean >= x
        lo, hi = 0, len(centroids)
        while lo < hi:
            mid = (lo + hi) // 2
            if centroids[mid].mean < x:
                lo = mid + 1
            else:
                hi = mid
        pos = lo

        # Compute cumulative weight before pos using a running sum
        # (we already have the centroids list)
        cum_weights = [0.0] * (len(centroids) + 1)
        for j in range(len(centroids)):
            cum_weights[j + 1] = cum_weights[j] + centroids[j].weight

        if pos == 0:
            c = centroids[0]
            inner_w = c.weight / 2.0
            frac = 1.0 if c.mean == self.min else (x - self.min) / (c.mean - self.min)
            return (inner_w * frac) / n

        if pos >= len(centroids):
            c = centroids[-1]
            cumulative = cum_weights[len(centroids) - 1]
            inner_w = c.weight / 2.0
            right_w = n - cumulative - c.weight / 2.0
            frac = 0.0 if self.max == c.mean else (x - c.mean) / (self.max - c.mean)
            return (cumulative + c.weight / 2.0 + right_w * frac) / n

        i = pos - 1
        c = centroids[i]
        next_c = centroids[pos]
        cumulative = cum_weights[i]

        if i == 0 and x < c.mean:
            inner_w = c.weight / 2.0
            frac = 1.0 if c.mean == self.min else (x - self.min) / (c.mean - self.min)
            return (inner_w * frac) / n
        if i == 0 and x == c.mean:
            return (c.weight / 2.0) / n

        if pos == len(centroids) - 1 and x > next_c.mean:
            next_cumulative = cumulative + c.weight
            inner_w = next_c.weight / 2.0
            right_w = n - next_cumulative - next_c.weight / 2.0
            frac = 0.0 if self.max == next_c.mean else (x - next_c.mean) / (self.max - next_c.mean)
            return (next_cumulative + next_c.weight / 2.0 + right_w * frac) / n

        next_cumulative = cumulative + c.weight
        mid_w = cumulative + c.weight / 2.0
        next_mid = next_cumulative + next_c.weight / 2.0

        if x < next_c.mean:
            if c.mean == next_c.mean:
                return (mid_w + (next_mid - mid_w) / 2.0) / n
            frac = (x - c.mean) / (next_c.mean - c.mean)
            return (mid_w + frac * (next_mid - mid_w)) / n

        return next_mid / n

    def centroid_count(self) -> int:
        if self._buffer:
            self.compress()
        return self._tree.size()

    # -- internals --------------------------------------------------------

    def _flush_for_merge(self) -> list[Centroid]:
        if self._buffer:
            self.compress()
        return self._tree.to_list()

    def _k(self, q: float) -> float:
        return (self.delta / (2.0 * math.pi)) * math.asin(2.0 * q - 1.0)

    @staticmethod
    def _merge_into_last(centroids: list[Centroid], c: Centroid) -> None:
        last = centroids[-1]
        new_weight = last.weight + c.weight
        last.mean = (last.mean * last.weight + c.mean * c.weight) / new_weight
        last.weight = new_weight

File: julia/TDigest.jl
"""
Dunning t-digest for online quantile estimation.
Merging digest variant with K_1 (arcsine) scale function.
Uses an array-backed 2-3-4 tree with monoidal measures.
Pure Julia -- no external packages.
"""
module TDigestModule

include("Tree234.jl")
using .Tree234Module

export TDigest, add!, compress!, quantile, cdf, merge!, centroid_count

struct Centroid
    mean::Float64
    weight::Float64
end

struct TdMeasure
    weight::Float64
    count::Int
    max_mean::Float64
    mean_weight_sum::Float64
end

const DEFAULT_DELTA = 100.0
const BUFFER_FACTOR = 5

# Trait functions for Tree234
function _td_measure(c::Centroid)::TdMeasure
    TdMeasure(c.weight, 1, c.mean, c.mean * c.weight)
end

function _td_combine(a::TdMeasure, b::TdMeasure)::TdMeasure
    TdMeasure(a.weight + b.weight, a.count + b.count,
              max(a.max_mean, b.max_mean),
              a.mean_weight_sum + b.mean_weight_sum)
end

function _td_identity()::TdMeasure
    TdMeasure(0.0, 0, -Inf, 0.0)
end

function _td_compare(a::Centroid, b::Centroid)::Int
    a.mean < b.mean ? -1 : (a.mean > b.mean ? 1 : 0)
end

mutable struct TDigest
    delta::Float64
    tree::Tree234{Centroid,TdMeasure}
    buffer::Vector{Centroid}
    total_weight::Float64
    min::Float64
    max::Float64
    buffer_cap::Int

    function TDigest(delta::Real = DEFAULT_DELTA)
        t = Tree234{Centroid,TdMeasure}(_td_measure, _td_combine,
                                         _td_identity, _td_compare)
        new(Float64(delta), t, Centroid[], 0.0, Inf, -Inf,
            ceil(Int, Float64(delta) * BUFFER_FACTOR))
    end
end

# -- Scale function -------------------------------------------------------

function _k(td::TDigest, q::Float64)::Float64
    (td.delta / (2.0 * pi)) * asin(2.0 * q - 1.0)
end

function _merge_centroid!(centroids::Vector{Centroid}, c::Centroid)
    last = centroids[end]
    new_weight = last.weight + c.weight
    new_mean = (last.mean * last.weight + c.mean * c.weight) / new_weight
    centroids[end] = Centroid(new_mean, new_weight)
end

# -- Mutation -------------------------------------------------------------

function add!(td::TDigest, value::Real, weight::Real = 1.0)
    v = Float64(value)
    w = Float64(weight)
    push!(td.buffer, Centroid(v, w))
    td.total_weight += w
    if v < td.min
        td.min = v
    end
    if v > td.max
        td.max = v
    end
    if length(td.buffer) >= td.buffer_cap
        compress!(td)
    end
    td
end

function compress!(td::TDigest)
    if isempty(td.buffer) && length(td.tree) <= 1
        return td
    end

    # Collect all centroids from tree and buffer
    all = collect_keys(td.tree)
    append!(all, td.buffer)
    empty!(td.buffer)
    sort!(all; by = c -> c.mean)

    new_centroids = Centroid[all[1]]
    weight_so_far = 0.0
    n = td.total_weight

    for i in 2:length(all)
        proposed = new_centroids[end].weight + all[i].weight
        q0 = weight_so_far / n
        q1 = (weight_so_far + proposed) / n

        if proposed <= 1 && length(all) > 1
            _merge_centroid!(new_centroids, all[i])
        elseif _k(td, q1) - _k(td, q0) <= 1.0
            _merge_centroid!(new_centroids, all[i])
        else
            weight_so_far += new_centroids[end].weight
            push!(new_centroids, Centroid(all[i].mean, all[i].weight))
        end
    end

    # Rebuild tree from sorted merged centroids
    build_from_sorted!(td.tree, new_centroids)
    td
end

# -- Queries --------------------------------------------------------------

function quantile(td::TDigest, q::Float64)
    if !isempty(td.buffer)
        compress!(td)
    end
    sz = length(td.tree)
    if sz == 0
        return nothing
    end

    centroids = collect_keys(td.tree)

    if sz == 1
        return centroids[1].mean
    end

    q = clamp(q, 0.0, 1.0)
    n = td.total_weight
    target = q * n
    nc = length(centroids)

    # Build prefix sums
    cum = Vector{Float64}(undef, nc)
    cum[1] = centroids[1].weight
    for i in 2:nc
        cum[i] = cum[i - 1] + centroids[i].weight
    end

    # Handle first centroid edge case
    first_c = centroids[1]
    if target < first_c.weight / 2.0
        if first_c.weight == 1.0
            return td.min
        end
        return td.min + (first_c.mean - td.min) * (target / (first_c.weight / 2.0))
    end

    # Handle last centroid edge case
    last_c = centroids[nc]
    if target > n - last_c.weight / 2.0
        if last_c.weight == 1.0
            return td.max
        end
        remaining = n - last_c.weight / 2.0
        return last_c.mean + (td.max - last_c.mean) * ((target - remaining) / (last_c.weight / 2.0))
    end

    # Binary search for position
    lo_idx = 1
    hi_idx = nc
    while lo_idx < hi_idx
        mid_idx = div(lo_idx + hi_idx, 2)
        if cum[mid_idx] < target
            lo_idx = mid_idx + 1
        else
            hi_idx = mid_idx
        end
    end
    idx = lo_idx

    if idx >= nc
        idx = nc - 1
    end
    if idx < 1
        idx = 1
    end

    cumulative = idx > 1 ? cum[idx - 1] : 0.0
    c = centroids[idx]
    mid_val = cumulative + c.weight / 2.0

    if idx > 1 && target < mid_val
        idx -= 1
        cumulative = idx > 1 ? cum[idx - 1] : 0.0
        c2 = centroids[idx]
        mid2 = cumulative + c2.weight / 2.0
        next_c = centroids[idx + 1]
        next_mid = cumulative + c2.weight + next_c.weight / 2.0
        frac = next_mid == mid2 ? 0.5 : (target - mid2) / (next_mid - mid2)
        return c2.mean + frac * (next_c.mean - c2.mean)
    end

    if idx == nc
        return c.mean
    end

    next_c = centroids[idx + 1]
    next_mid = cumulative + c.weight + next_c.weight / 2.0

    if target <= next_mid
        frac = next_mid == mid_val ? 0.5 : (target - mid_val) / (next_mid - mid_val)
        return c.mean + frac * (next_c.mean - c.mean)
    end

    td.max
end

function cdf(td::TDigest, x::Real)
    x = Float64(x)
    if !isempty(td.buffer)
        compress!(td)
    end
    sz = length(td.tree)
    if sz == 0
        return nothing
    end
    if x <= td.min
        return 0.0
    end
    if x >= td.max
        return 1.0
    end

    centroids = collect_keys(td.tree)
    nc = length(centroids)
    n = td.total_weight

    # Build prefix sums
    cum = Vector{Float64}(undef, nc)
    cum[1] = centroids[1].weight
    for i in 2:nc
        cum[i] = cum[i - 1] + centroids[i].weight
    end

    # Binary search for position: find first centroid with mean >= x
    means = [c.mean for c in centroids]
    pos = searchsortedfirst(means, x)

    # x is less than the first centroid's mean
    if pos == 1
        c = centroids[1]
        if x < c.mean
            inner_w = c.weight / 2.0
            frac = c.mean == td.min ? 1.0 : (x - td.min) / (c.mean - td.min)
            return (inner_w * frac) / n
        end
        return (c.weight / 2.0) / n
    end

    # x >= all centroid means
    if pos > nc
        c = centroids[nc]
        cumulative = nc > 1 ? cum[nc - 1] : 0.0
        if x > c.mean
            right_w = n - cumulative - c.weight / 2.0
            frac = td.max == c.mean ? 0.0 : (x - c.mean) / (td.max - c.mean)
            return (cumulative + c.weight / 2.0 + right_w * frac) / n
        end
        return (cumulative + c.weight / 2.0) / n
    end

    # x is between centroids[pos-1].mean and centroids[pos].mean
    i = pos - 1
    c = centroids[i]
    next_c = centroids[pos]

    cumulative = i > 1 ? cum[i - 1] : 0.0
    mid_cdf = cumulative + c.weight / 2.0
    next_cumulative = cumulative + c.weight
    next_mid = next_cumulative + next_c.weight / 2.0

    if i == nc
        if x > c.mean
            right_w = n - cumulative - c.weight / 2.0
            frac = td.max == c.mean ? 0.0 : (x - c.mean) / (td.max - c.mean)
            return (cumulative + c.weight / 2.0 + right_w * frac) / n
        end
        return (cumulative + c.weight / 2.0) / n
    end

    if x < next_c.mean
        if c.mean == next_c.mean
            return (mid_cdf + (next_mid - mid_cdf) / 2.0) / n
        end
        frac = (x - c.mean) / (next_c.mean - c.mean)
        return (mid_cdf + frac * (next_mid - mid_cdf)) / n
    end

    return next_mid / n
end

function merge!(td::TDigest, other::TDigest)
    if !isempty(other.buffer)
        compress!(other)
    end
    other_centroids = collect_keys(other.tree)
    for c in other_centroids
        add!(td, c.mean, c.weight)
    end
    td
end

function centroid_count(td::TDigest)::Int
    if !isempty(td.buffer)
        compress!(td)
    end
    length(td.tree)
end

end # module
